mifospay/src/main/java/org/mifospay/savedcards/ui/AddCardDialogSheet.kt

```diff
@@ -0,0 +1,178 @@
+        Row(modifier = Modifier.fillMaxWidth()) {
+            MifosOutlinedTextField(
+                value = expiration,
+                onValueChange = { if (it.length <= 4) expiration = it },
+                modifier = Modifier.weight(1f),
+                label = R.string.mm_yy,
+                visualTransformation = ExpirationDateMask(),
+                keyboardOptions = KeyboardOptions.Default.copy(keyboardType = KeyboardType.NumberPassword)
+            )
+            Spacer(modifier = Modifier.width(16.dp))
+            MifosOutlinedTextField(
+                value = cvv,
+                onValueChange = { if (it.length <= 3) cvv = it },
+                modifier = Modifier.weight(1f),
+                label = R.string.cvv,
+                keyboardOptions = KeyboardOptions.Default.copy(keyboardType = KeyboardType.NumberPassword)
+            )
+        }
+        Spacer(modifier = Modifier.height(16.dp))
+        Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
+            MifosButton(onClick = { cancelClicked() }) {
+                Text(text = stringResource(id = R.string.cancel))
+            }
+            Spacer(modifier = Modifier.width(16.dp))
+            MifosButton(onClick = {
+                val card = Card(
+                    cardNumber = creditCardNumber,
+                    cvv = cvv,
+                    expiryDate = expiration,
+                    firstName = firstName,
+                    lastName = lastName
+                )
+
+                validateCard(card, onError = {
+                    Toast.makeText(context, context.getString(it), Toast.LENGTH_SHORT).show()
+                }, onSuccess = { addClicked(it) })
+            }) {
+                Text(text = stringResource(id = R.string.add))
+            }
+        }
+    }
+}
+
+fun validateCard(card: Card, onError: (Int) -> Unit, onSuccess: (Card) -> Unit) {
+
+    when {
+        card.cardNumber.isEmpty() || card.cvv.isEmpty() || card.expiryDate.isEmpty() || card.firstName.isEmpty() || card.lastName.isEmpty() -> {
+            onError(R.string.all_fields_required)
+        }
+
+        card.cardNumber.length < 16 || !CreditCardUtils.validateCreditCardNumber(card.cardNumber) -> {
+            onError(R.string.invalid_credit_card)
+        }
+
+        card.expiryDate.length < 4 -> {
+            onError(R.string.expiry_date_length_error)
+        }
+
+        (card.expiryDate.substring(2, 4) == Calendar.getInstance()[Calendar.YEAR].toString()
+            .substring(2, 4) && card.expiryDate.substring(0, 2)
+            .toInt() < Calendar.getInstance()[Calendar.MONTH] + 1) || card.expiryDate.substring(
+            0,
+            2
+        ).toInt() > 12 -> {
+            onError(R.string.invalid_expiry_date)
+        }
+
+        card.cvv.length < 3 -> {
+            onError(R.string.cvv_length_error)
+        }
+
+        else -> {
+            onSuccess(card)
+        }
+    }
```

Comment:
have this validation in the `AddCardDialogSheetContent` itself, like following: 
    
    var amount by rememberSaveable { mutableStateOf(prefilledAmount) }
    var amountValidator by rememberSaveable { mutableStateOf<String?>(null) }
    
    LaunchedEffect(key1 = amount) {
        amountValidator = when {
            amount.trim().isEmpty() -> context.getString(R.string.enter_amount)
            amount.trim().toDoubleOrNull() == null -> context.getString(R.string.enter_valid_amount)
            amount.trim().toDouble().compareTo(0.0) <= 0 -> context.getString(R.string.enter_valid_amount)
            else -> null
        }
    }